### PR TITLE
Switch to ESLint instead of Standard JS

### PR DESCRIPTION
**Pull request by Nasicus**
## Why
Standard JS is fine and good (ok actually I don't like it, but that's another matter 😆 ) but in my opinion some important stuff is missing, for example: Warnings or errors about using `var`.
Ultimately ESLint offers us **flexibility** and if we ever want to move away from standard js, we could do so without making a fuss.

(Second reason: Somehow my editor (webstorm) is too stupid to work with the standard js rules for `globals` like `angular` and the ignored directories / files)

## Do we need to change everything?
No! ESLint can extend the "standard" rules, so it's almost the same!
Actually with my current commit, we "only" have 73 errors and all of them are somehow because the indention is not correct (I guess ESLint has a slight difference there sometimes)

## Additional Rules (if we want to use ESLint)
Rules which we should enable (based on review comments I saw of the ark team)
- [x]  `no-var` [rule](https://eslint.org/docs/rules/no-var)
  - I don't know if @j-a-m-l  has already done something here? You mentioned something to me!
- [x] `prefer-arrow-callback` [rule](https://eslint.org/docs/rules/prefer-arrow-callback)
- ???

## ToDo's (if we want to use ESLint)
- [x] In my opinon we should NOT exclude the  `test` directory from the linting => so basically include it and fix the errors there
- [x] Adjust `travis` build server so he executes the `eslint` npm task
- ???

## The killer argument against it?!?

We probably need to remove the standard js badge:
badge
Not sure if [certain](https://github.com/ArkEcosystem/ark-desktop/pull/376#issuecomment-341390409) people could live with this!

### CODE REVIEW

**Comment by luciorubeens** [comment]
This discussion has already been pull a few times, I agree with you that some rules are missing, so this proposal is valid..

Related:
https://github.com/ArkEcosystem/ark-desktop/issues/323
https://github.com/ArkEcosystem/ark-desktop/pull/376

**Comment by j-a-m-l** [comment]
I don't care about the badge (it was a joke).

My only concern is that Standard JS solves 1 problem that it has been in JS for a while: creates an standard way to format JS code accepted by different communities and large projects. If we move to ESLint we open a Pandora's box: everybody would be able to add their own tastes instead of conforming to [one](https://www.youtube.com/watch?v=MqY83YjA0JE&feature=youtu.be&t=01m36s).

Anyway, if you prefer that, I wouldn't complain.

**Comment by Nasicus** [comment]
I know it was a joke, I just tried (and apparently failed) to contimue it ;)

Personally I find it more important that the code style is constent within the project and that we (you) are happy with how it is.
Since you guys often mention "don't use var" and/or use "arrow functions" I find the additional rules (which are not standard anymore) are appropiate :)

I'll give this a go then!

**Comment by Nasicus** [comment]
Ok added ESLint everywhere (including tests) and added the rules `no-var` and `prefer-arrow-callback` and fixed all files.

Some notes:

`karma.conf.js`:
 - `_filenameOverride` for some reason, if I use  `path.join` some tests will fail. It appears the path for the other requires is then wrong... do not really understand why...also it appearsthe filename `index.html` is irrelevant... but if I leave it as it is (with a comment to disable eslint rule there) it works
- I removed `coverage` from the reports list, because it takes so much space away in the test window and you always have to scroll up to see the errors / results

By the way:  Review / merge the other big PRs (e.g. voter refactoring) first, because else they will have a LOT of conflicts. It's probably easier (and fairer) if I fix the conflicts in my branch.

**Comment by luciorubeens** [comment]
@alexbarnsley Give us your 2 cent again 😄

**Comment by Nasicus** [comment]
So.... we all know travis hates me...but still for him to do that?!?

The problem is, that travis build does NOT fail and I don't understand why.
In my last commit I checked in files which are formatted wrong, but on travis `eslint` always succeeds....does anyone have an idea what could be wrong?

**Comment by zillionn** [comment]
Try
"lint": "eslint **/*.js"
and
npm run lint

**Comment by Nasicus** [comment]
Ok I'm sooooooooooooo 😠  and travis is a piece of 💩 

So as I've written above travis always suceeded, _instead_ of _failing_ for eslint.
This was apparently my own mistake, since I needed to write the npm command like this `"eslint \"**/*.js\""` instead of `"eslint **/*.js"`.

I can live with that. I can stand by mistakes, I can fix them.

But then, when travis _finally started failing_, he didn't want to stop anymore and I even as I reverted my on purpose checked in mistakes, he still gave me this:

```
/home/travis/build/ArkEcosystem/ark-desktop/client/app/app.js
  22:1  error  Expected indentation of 4 spaces but found 2  indent
  23:1  error  Expected indentation of 4 spaces but found 2  indent
  24:1  error  Expected indentation of 4 spaces but found 2  indent
  25:1  error  Expected indentation of 4 spaces but found 2  indent
 ```

But locally it still worked ? **WHAT**? Ok..ok, I told to myself, take it easy, just breathe, be smart, try it on your Ubuntu, maybe it has something to do with the line endings or something similar, and so I did. And guess what? There it worked out of the box (freshly cloned, freshly `npm installed`). **WHAT**? **WHAT**? **WHAT**?
Ok ok...back to zero...I had the glorious idea that travis maybe not always checks out all files or mixes something up, so I switchted to `master` and merged it back into my branch...and _WHAT DID I SEE. THERE_!?! The spaces were really changed (thanks @luciorubeens ) ! I fixed them, [pushed again](https://www.youtube.com/watch?v=d2E_iiTd9hQ), and IT FINALLY WORKS.

TLDR;
Travis is stupid. Even though MY branch did NOT have the `app.js` with the changed spaces, it somehow JUST took it from the `master` branch? Seriously? That's just fucking wrong.

Anyway it works now.
I'm Out. (For now.) 

Thanks for listening. Next story will not be free, a fee of 1 ARK per read, will be requested.

PS: @luciorubeens no hard feelings - just kidding ❤️

**Comment by luciorubeens** [comment]
Hey @Nasicus, great history man 🤣 

```bash
$ npm run eslint

> eslint "**/*.js"

/Users/lucio/Ark/ark-desktop/.eslintrc.json:
	Configuration for rule "indent" is invalid:
	Value "off" no (or more than one) schemas match.
	Value "off" no (or more than one) schemas match.

Error: /Users/lucio/Ark/ark-desktop/.eslintrc.json:
	Configuration for rule "indent" is invalid:
	Value "off" no (or more than one) schemas match.
	Value "off" no (or more than one) schemas match.
```

**Comment by Nasicus** [comment]
Strange tested it on friday and it worked. Did you run `npm install` ? Else I will look at it on sunday.

**Comment by Nasicus** [comment]
Strange tested it on friday and it worked. Did you run `npm install` ? Else I will look at it on sunday.

**Comment by Nasicus** [comment]
Strange tested it on friday and it worked. Did you run `npm install` ? Else I will look at it on sunday.

**Comment by luciorubeens** [comment]
Oh sorry, my fault. Need to upgrade eslint

**Comment by luciorubeens** [comment]
Finally your PR has been merged, good job! 👏
